File: backend/python/predict.py
"""
Predict cancer type from a mutation profile.
Reads JSON from stdin: {"mutations": ["TP53_R175H", "BRCA1_185delAG"]}
"""

import sys
import os
import json
import warnings
import joblib
import pandas as pd

warnings.filterwarnings("ignore")


def main():
    # Read from stdin
    input_text = sys.stdin.read().strip()
    if not input_text:
        print(json.dumps({"error": "No input provided"}))
        sys.exit(1)

    model_dir = os.path.join(os.path.dirname(__file__), "..", "model")
    model_path = os.path.join(model_dir, "cancer_model.joblib")
    le_path = os.path.join(model_dir, "label_encoder.joblib")
    meta_path = os.path.join(model_dir, "metadata.json")

    if not os.path.exists(model_path):
        print(json.dumps({"error": "Model not found. Run train_model.py first."}))
        sys.exit(1)

    model = joblib.load(model_path)
    le = joblib.load(le_path)
    with open(meta_path) as f:
        metadata = json.load(f)

    features = metadata["features"]
    input_data = json.loads(input_text)
    mutations = input_data.get("mutations", [])

    # Build binary feature vector
    feature_vector = [1 if f in mutations else 0 for f in features]
    X = pd.DataFrame([feature_vector], columns=features)

    probabilities = model.predict_proba(X)[0]
    cancer_types = le.classes_

    # Build sorted predictions
    predictions = []
    for i, cancer in enumerate(cancer_types):
        predictions.append({
            "cancer_type": cancer,
            "probability": round(float(probabilities[i]) * 100, 1),
        })
    predictions.sort(key=lambda x: x["probability"], reverse=True)

    top = predictions[0]

    print(json.dumps({
        "top_prediction": top["cancer_type"],
        "top_probability": top["probability"],
        "all_predictions": predictions,
        "mutations_analyzed": mutations,
        "model_accuracy": metadata["accuracy"],
    }))


if __name__ == "__main__":
    main()

File: backend/routes/predict.js
const path = require("path");
const fs = require("fs");

let rfModel = null;
let metadata = null;

function loadModel() {
  if (rfModel && metadata) return;
  const modelPath = path.join(__dirname, "..", "model", "rf_model.json");
  const metaPath = path.join(__dirname, "..", "model", "metadata.json");

  try {
    if (fs.existsSync(modelPath) && fs.existsSync(metaPath)) {
      rfModel = JSON.parse(fs.readFileSync(modelPath, "utf-8"));
      metadata = JSON.parse(fs.readFileSync(metaPath, "utf-8"));
    }
  } catch (err) {
    console.error("Failed to parse model JSON:", err);
  }
}

function predictProbabilities(featureVector) {
  const nClasses = rfModel.n_classes;
  const nTrees = rfModel.trees.length;

  let classProbs = new Array(nClasses).fill(0);

  for (const tree of rfModel.trees) {
    let node = 0;
    while (tree.children_left[node] !== -1) {
      const f = tree.feature[node];
      const t = tree.threshold[node];
      if (featureVector[f] <= t) {
        node = tree.children_left[node];
      } else {
        node = tree.children_right[node];
      }
    }

    // In scikit-learn, values are shape (n_nodes, 1, n_classes)
    const values = tree.value[node][0];
    const sum = values.reduce((a, b) => a + b, 0);
    // Protect against division by zero
    if (sum > 0) {
      for (let c = 0; c < nClasses; c++) {
        classProbs[c] += values[c] / sum;
      }
    }
  }

  // Average across all trees
  for (let c = 0; c < nClasses; c++) {
    classProbs[c] /= nTrees;
  }
  return classProbs;
}

async function predictRoutes(fastify) {
  const metadataPath = path.join(__dirname, "..", "model", "metadata.json");

  fastify.post("/predict", async (request, reply) => {
    let { mutations } = request.body || {};

    // Fallback if mutations array is missing or invalid
    if (!Array.isArray(mutations)) {
      mutations = [];
    }

    // Default fallback prediction in case anything goes wrong
    // This ensures the frontend never crashes and always gets a valid response format.
    const defaultPrediction = {
      cancer_type: "Unknown",
      probability: 0
    };

    const fallbackResponse = {
      top_prediction: defaultPrediction.cancer_type,
      top_probability: defaultPrediction.probability,
      all_predictions: [defaultPrediction],
      mutations_analyzed: mutations,
      model_accuracy: 0,
      warning: "Model unavailable or prediction failed. Returning fallback data."
    };

    try {
      loadModel();

      // If model failed to load, return fallback gracefully
      if (!rfModel || !metadata || !metadata.features || !metadata.cancer_types) {
        return reply.send(fallbackResponse);
      }

      const featureVector = metadata.features.map(f => mutations.includes(f) ? 1 : 0);
      const probabilities = predictProbabilities(featureVector);

      let predictions = [];
      const cancerTypes = metadata.cancer_types;
      for (let i = 0; i < cancerTypes.length; i++) {
        predictions.push({
          cancer_type: cancerTypes[i],
          probability: Math.round(probabilities[i] * 1000) / 10
        });
      }

      predictions.sort((a, b) => b.probability - a.probability);

      const top = predictions[0] || defaultPrediction;

      return reply.send({
        top_prediction: top.cancer_type,
        top_probability: top.probability,
        all_predictions: predictions,
        mutations_analyzed: mutations,
        model_accuracy: metadata.accuracy || 0
      });
    } catch (err) {
      fastify.log.error("Prediction Error:", err);
      // Failsafe: return the fallback response instead of a 500 status code
      return reply.send(fallbackResponse);
    }
  });

  fastify.get("/model-info", async (request, reply) => {
    try {
      if (!fs.existsSync(metadataPath)) {
        return reply.send({ error: "Model not trained yet." }); // Return 200 with error property
      }
      return reply.send(JSON.parse(fs.readFileSync(metadataPath, "utf-8")));
    } catch (err) {
      return reply.send({ error: "Failed to read model info." });
    }
  });
}

module.exports = predictRoutes;
